### PR TITLE
[FLINK-12283][table] Relax UDF constraints by using the ClosureCleaner

### PR DESCRIPTION
**Pull request by twalthr**
## What is the purpose of the change

This PR relaxes the UDF constraints as much as possible by using the ClosureCleaner. For Java users, this is very convenient. For Scala users, it improves the current status but we still need to fix FLINK-15162.

## Brief change log

This PR ensures that the validation and cleaning happens at exactly 2 locations. Either during registration in the function catalog or during resolution of inline, unregistered functions.

Since the ClosureCleaner needs configuration parameters, this PR also adds configuration through the stack. But this makes sense for the future in any case because configuration should be the first known instance when creating a table environment.

## Verifying this change

Integration tests added for Blink and legacy planner.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? JavaDocs

### CODE REVIEW

**Comment by twalthr** [comment]
@zjffdu does this PR also works for you? Or will the ClosureCleaner complain about the non-serializability?

**Comment by zjffdu** [comment]
@twalthr Could you do a rebase against master so that I can test is in Zeppelin ?

**Comment by twalthr** [comment]
@zjffdu rebased

**Comment by zjffdu** [comment]
It works in Zeppelin, thanks @twalthr  +1

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/expressions/resolver/rules/ResolveCallByArgumentsRule.java [hunk 1]
```diff
@@ -220,6 +227,43 @@ private ResolvedExpression runLegacyTypeInference(
...
 		}
+
+		/**
+		 * Validates and cleans an inline, unregistered {@link UserDefinedFunction}.
+		 */
+		private FunctionDefinition prepareUserDefinedFunction(FunctionDefinition definition) {
+			if (definition instanceof ScalarFunctionDefinition) {
+				final ScalarFunctionDefinition sf = (ScalarFunctionDefinition) definition;
+				UserDefinedFunctionHelper.prepareFunction(resolutionContext.configuration(), sf.getScalarFunction());
+				return new ScalarFunctionDefinition(
```
Comment (aljoscha) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/expressions/resolver/rules/ResolveCallByArgumentsRule.java]: Why do we need to return new definitions from this methdod? Having the if blocks here can break things in the future if we add more function definition types.

**Comment by aljoscha** [comment]
I had one inline comment. Other than that this PR seems good!

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/expressions/resolver/rules/ResolveCallByArgumentsRule.java (hunk 1 above)
Comment (twalthr) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/expressions/resolver/rules/ResolveCallByArgumentsRule.java]: This method is just a temporary solution until FLIP-65 is implemented. In the future, we don't need to store the type next to the function but the function provides a strategy for determining the type.

Comment (aljoscha) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/expressions/resolver/rules/ResolveCallByArgumentsRule.java]: Ok, good to merge, then!

**Comment by twalthr** [comment]
Thank you @aljoscha. Merging...

**Comment by twalthr** [comment]
@flinkbot run travis

**Pull request closed without merging**